### PR TITLE
Expose SchemaFactoryAlpha as an alpha API

### PR DESCRIPTION
**Pull request by Abe27342**
## Description

Exposes `SchemaFactoryAlpha` and associated additional options on `object` and `objectRecursive` to tree's alpha API surface. This lights up the feature implemented in #23192, so I have also added a changeset with some user-facing guidance on that feature.

### CODE REVIEW

File: packages/dds/tree/src/simple-tree/api/schemaFactoryAlpha.ts [hunk 1]
```diff
@@ -3,37 +3,34 @@
...
- * Copy of {@link SchemaFactory} with additional alpha APIs.
+ * {@link SchemaFactory} with additional alpha APIs.
  *
+ * @alpha
  * @privateRemarks
- * Not currently exported to the public API surface as doing so produces errors in API-extractor.
  *
- * Can be removed once additional object node features are deemed stable and on the base class.
+ * Defining this class separately allows correct API isolation of alpha APIs (method signatures) on SchemaFactory.
+ * Prefer defining duplicate private methods in `SchemaFactoryAlpha` rather than increasing their exposure, as `protected`
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/schemaFactoryAlpha.ts:27]: Nit: it took me a couple of re-reads to understand what you meant here. Might be worth attempting to reword.

File: .changeset/moody-pandas-fall.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
+---
```
Comment (Josmithr) [inline .changeset/moody-pandas-fall.md:1]: Those poor moody pandas 😢

File: .changeset/moody-pandas-fall.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	x: factory.number,
+	y: factory.number,
+	// Note that radius and color must both be declared as optional fields since this application must
+	// support opening up existing documents that didn't have this information.
+	radius: factory.optional(factory.number),
+	color: factory.optional(factory.string), // ex: #00FF00
+}, { allowUnknownOptionalFields: true }) {}
+```
+
+When they go to deploy this newer version of the application, they could opt to start upgrading documents as soon as the newer code rolled out and the older code would still be able to open up (and collaborate on) documents using the newer schema version.
```
Comment (Josmithr) [inline .changeset/moody-pandas-fall.md:50]: Nit:
```suggestion
When they go to deploy this newer version of the application, they could opt to start upgrading documents as soon as the newer code is rolled out, and the older code would still be able to open up (and collaborate on) documents using the newer schema version.
```

File: packages/dds/tree/src/simple-tree/api/schemaFactory.ts [hunk 4]
```diff
@@ -103,6 +103,8 @@ export function schemaFromValue(value: TreeValue): TreeNodeSchema {
 
 /**
  * Options when declaring an {@link SchemaFactory.object|object node}'s schema
+ *
+ * @alpha
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/schemaFactory.ts:107]: Nit: Question about the documentation on the property below. The first line reads:
> Opt in to viewing documents which have unknown optional fields for this object's identifier, i.e. optional fields
> in the document schema which are not present in this object schema declaration.

I'm not sure what "identifier" means in this context. Should this instead just say something like:
> Allow nodes typed with this object node schema to contain optional fields that are not present in the schema declaration.

?

File: packages/dds/tree/api-report/tree.legacy.public.api.md [hunk 5]
```diff
@@ -375,7 +375,7 @@ export class SchemaFactory<out TScope extends string | undefined = string | unde
...
-    objectRecursive<const Name extends TName, const T extends Unenforced<RestrictiveStringRecord<ImplicitFieldSchema>>>(name: Name, t: T): TreeNodeSchemaClass<ScopedSchemaName<TScope, Name>, NodeKind.Object, TreeObjectNodeUnsafe<T, ScopedSchemaName<TScope, Name>>, object & { readonly [Property in keyof T as FieldHasDefaultUnsafe<T[Property]> extends false ? Property : never]: InsertableTreeFieldFromImplicitFieldUnsafe<T[Property], UnionToIntersection_2<T[Property]>>; } & { readonly [Property_1 in keyof T as FieldHasDefaultUnsafe<T[Property_1]> extends true ? Property_1 : never]?: InsertableTreeFieldFromImplicitFieldUnsafe<T[Property_1], UnionToIntersection_2<T[Property_1]>> | undefined; }, false, T>;
+    objectRecursive<const Name extends TName, const T extends Unenforced<RestrictiveStringRecord<ImplicitFieldSchema>>>(name: Name, t: T): TreeNodeSchemaClass<ScopedSchemaName<TScope, Name>, NodeKind.Object, TreeObjectNodeUnsafe<T, ScopedSchemaName<TScope, Name>>, object & InsertableObjectFromSchemaRecordUnsafe<T>, false, T>;
```
Comment (Josmithr) [inline packages/dds/tree/api-report/tree.legacy.public.api.md:378]: Sanity check: is this expected?

**Review by Josmithr — APPROVED** [review]
Left a few suggestions but otherwise looks great to me! And lines up very well with unrelated schema factory functionality I'm working on.

You mentioned that API-Extractor was complaining about these changes earlier, but no longer seems to be? Let me know if I can help investigate anything there. But assuming it continues to be stable, looks good to me!

File: packages/dds/tree/api-report/tree.legacy.public.api.md (hunk 5 above)
Comment (Abe27342) [reply packages/dds/tree/api-report/tree.legacy.public.api.md:378]: It seems reasonable for me for it to change. The "before" appears to be the expanded form of the "after", and the change is reasonably explained by type emission being slightly different due to explicitly specifying the return type of `objectRecursive`.

**Comment by Abe27342** [comment]
> Left a few suggestions but otherwise looks great to me! And lines up very well with unrelated schema factory functionality I'm working on.
> 
> You mentioned that API-Extractor was complaining about these changes earlier, but no longer seems to be? Let me know if I can help investigate anything there. But assuming it continues to be stable, looks good to me!

Yeah, I'm not totally sure it was actually an API-extractor issue, it just surfaced that way. I think the original problem I was seeing was typescript emitting 'any' which down the line caused problems, but it's plausible that when I last tried this I had missed something subtle (previous PR had a lot of changes) and when reimplementing this from a fresh state I didn't make the same mistake.

File: packages/dds/tree/src/simple-tree/api/schemaFactory.ts (hunk 4 above)
Comment (Abe27342) [reply packages/dds/tree/src/simple-tree/api/schemaFactory.ts:107]: I like that way of putting it. `identifier` refers to the schema's identifier, but it's not particularly necessary to mention that here.

**Pull request merged**